### PR TITLE
DateBox - Support IME input for the masks (T905190)

### PR DESCRIPTION
**Pull request by alexander-kotov-dx**

### CODE REVIEW

**Comment by San4es** [comment]
https://github.com/DevExpress/DevExtreme/pull/14087/checks?check_run_id=940835186

File: js/ui/date_box/ui.date_box.mask.js [hunk 1]
```diff
@@ -129,23 +130,37 @@ const DateBoxMask = DateBoxBase.inherit({
         return typeof key === 'string' && key.length === 1 && !e.ctrl && !e.alt;
     },
 
+    isSingleDigitKey(e) {
```
Comment (San4es) [inline js/ui/date_box/ui.date_box.mask.js]:
```suggestion
    _isSingleDigitKey(e) {
```

File: js/ui/date_box/ui.date_box.mask.js [hunk 2]
```diff
@@ -129,23 +130,37 @@ const DateBoxMask = DateBoxBase.inherit({
...
+        const data = e.originalEvent?.data;
+        return data?.length === 1 && parseInt(data);
```
Comment (San4es) [inline js/ui/date_box/ui.date_box.mask.js]:
```suggestion
        return data?.length === 1 && parseInt(data, 10);
```

File: js/ui/date_box/ui.date_box.mask.js [hunk 3]
```diff
@@ -129,23 +130,37 @@ const DateBoxMask = DateBoxBase.inherit({
...
     _keyboardHandler(e) {
-        const key = e.originalEvent.key;
+        let key = e.originalEvent.key;
 
         const result = this.callBase(e);
 
-        if(!this._useMaskBehavior() || this._useBeforeInputEvent() || !this._isSingleCharKey(e)) {
-            return result;
-        }
+        if((browser.chrome) && e.key === 'Process' && e.code.indexOf('Digit') === 0) {
```
Comment (San4es) [inline js/ui/date_box/ui.date_box.mask.js:152]: Why we do not check `this._useMaskBehavior()` and `this._useBeforeInputEvent()` here?

File: js/ui/date_box/ui.date_box.mask.js [hunk 4]
```diff
@@ -356,15 +375,21 @@ const DateBoxMask = DateBoxBase.inherit({
         eventsEngine.on(this._input(), addNamespace('dxclick', MASK_EVENT_NAMESPACE), this._maskClickHandler.bind(this));
         eventsEngine.on(this._input(), addNamespace('paste', MASK_EVENT_NAMESPACE), this._maskPasteHandler.bind(this));
         eventsEngine.on(this._input(), addNamespace('drop', MASK_EVENT_NAMESPACE), () => {
-            this._renderDisplayText(this._getDisplayedText(this._maskValue));
-            this._selectNextPart();
+            this._renderSelectedPart();
         });
 
+        eventsEngine.on(this._input(), 'compositionend', this._maskCompositionEndHandler.bind(this));
```
Comment (San4es) [inline js/ui/date_box/ui.date_box.mask.js]: Where we unsubscribe this event handler?

```suggestion
        eventsEngine.on(this._input(), addNamespace('compositionend', MASK_EVENT_NAMESPACE), this._maskCompositionEndHandler.bind(this));
```

File: js/ui/date_box/ui.date_box.mask.js [hunk 5]
```diff
@@ -183,6 +197,11 @@ const DateBoxMask = DateBoxBase.inherit({
     },
 
     _keyPressHandler(e) {
+        if(e.originalEvent && e.originalEvent.inputType === 'insertCompositionText' && this.isSingleDigitKey(e)) {
```
Comment (DokaRus) [inline js/ui/date_box/ui.date_box.mask.js]:
```suggestion
        const { originalEvent: event } = e;
        if(event?.inputType === 'insertCompositionText' && this.isSingleDigitKey(e)) {
```

File: js/ui/date_box/ui.date_box.mask.js [hunk 6]
```diff
@@ -183,6 +197,11 @@ const DateBoxMask = DateBoxBase.inherit({
...
+            this._processInputKey(e.originalEvent.data);
```
Comment (DokaRus) [inline js/ui/date_box/ui.date_box.mask.js]:
```suggestion
            this._processInputKey(event.data);
```

Comment (alexander-kotov-dx) [reply js/ui/date_box/ui.date_box.mask.js:201]: 'event.data' does not work in FF

**Review by DokaRus — APPROVED** [review]

**Review by San4es — APPROVED** [review]
LGTM

**Pull request merged**